### PR TITLE
[roboteam_ai] Change of tactics

### PR DESCRIPTION
**Pull request by JornJorn**
Hungarian algorithm for deciding which pass defender goes to which robot

Pass defender used to switch around constantly, like https://drive.google.com/file/d/1KFf8YVQ_ehoMB_BYDZE2HUzlb_9aQyLl/view?usp=share_link
Now they don't, https://drive.google.com/file/d/1aplwvQSKXQj0PXLd_BRz9MIOUMMEju5m/view?usp=share_link

### CODE REVIEW

File: roboteam_ai/src/stp/plays/offensive/ChippingPass.cpp [hunk 1]
```diff
@@ -178,6 +150,73 @@ void ChippingPass::calculateInfoForAttackers() noexcept {
...
+        enemyMap.insert({score, enemy->getPos()});
+    }
+
+    if (enemyMap.size() < 3) {
+        stpInfos["pass_defender_1"].setPositionToDefend(
+            Vector2{field.middleLeftGrid.getOffSetY() + field.middleLeftGrid.getRegionHeight() / 2, field.middleLeftGrid.getOffSetX() + field.middleLeftGrid.getRegionWidth() / 2});
+        stpInfos["pass_defender_2"].setPositionToDefend(
+            Vector2{field.middleMidGrid.getOffSetY() + field.middleMidGrid.getRegionHeight() / 2, field.middleMidGrid.getOffSetX() + field.middleMidGrid.getRegionWidth() / 2});
+        stpInfos["pass_defender_3"].setPositionToDefend(Vector2{field.middleRightGrid.getOffSetY() + field.middleRightGrid.getRegionHeight() / 2,
+                                                                field.middleRightGrid.getOffSetX() + field.middleRightGrid.getRegionWidth() / 2});
```
Comment (emielsteerneman) [inline roboteam_ai/src/stp/plays/offensive/ChippingPass.cpp:174]: Quite difficult to read. Maybe split this up into more lines

**Review by emielsteerneman — DISMISSED** [review]
Looks functionally correct. A few comments:

* Does this work for for any amount of robots (both ours or opponents)? 
* Could use more comments in the code
* Could use some readability improvements, although I see how that's subjective

Please split up changes into multiple commits. Over half of these changes were good ones, but had nothing to do with `Improve pass defender location`. This makes it harder to review since one has to check all files for any functional changes related to `Improve pass defender location`

File: roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp [hunk 2]
```diff
@@ -180,6 +152,73 @@ void AttackingPass::calculateInfoForAttackers() noexcept {
...
+        stpInfos["pass_defender_1"].setPositionToDefend(
+            Vector2{field.middleLeftGrid.getOffSetY() + field.middleLeftGrid.getRegionHeight() / 2, field.middleLeftGrid.getOffSetX() + field.middleLeftGrid.getRegionWidth() / 2});
+        stpInfos["pass_defender_2"].setPositionToDefend(
+            Vector2{field.middleMidGrid.getOffSetY() + field.middleMidGrid.getRegionHeight() / 2, field.middleMidGrid.getOffSetX() + field.middleMidGrid.getRegionWidth() / 2});
+        stpInfos["pass_defender_3"].setPositionToDefend(Vector2{field.middleRightGrid.getOffSetY() + field.middleRightGrid.getRegionHeight() / 2,
+                                                                field.middleRightGrid.getOffSetX() + field.middleRightGrid.getRegionWidth() / 2});
+    } else {
+        Vector2 enemy1 = enemyMap.begin()->second;
+        stpInfos["pass_defender_1"].setPositionToDefend(enemy1);
+        enemyMap.erase(enemyMap.begin());
```
[the remark this answers is not in the document]
Comment (JornJorn) [reply roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp]: Yep that should be fine

File: roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp [hunk 3]
```diff
@@ -180,6 +152,73 @@ void AttackingPass::calculateInfoForAttackers() noexcept {
...
+        Vector2 enemy2 = enemyMap.begin()->second;
+        stpInfos["pass_defender_2"].setPositionToDefend(enemy2);
+        enemyMap.erase(enemyMap.begin());
+        Vector2 enemy3 = enemyMap.begin()->second;
+        stpInfos["pass_defender_3"].setPositionToDefend(enemy3);
```
[the remark this answers is not in the document]
Comment (JornJorn) [reply roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp]: Fixed for now, see ```calculateInfoForDefenders``` in ```roboteam_ai/src/stp/computations/PositionComputations.cpp```

File: roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp [hunk 4]
```diff
@@ -180,6 +152,73 @@ void AttackingPass::calculateInfoForAttackers() noexcept {
...
+        stpInfos["pass_defender_3"].setPositionToDefend(Vector2{field.middleRightGrid.getOffSetY() + field.middleRightGrid.getRegionHeight() / 2,
+                                                                field.middleRightGrid.getOffSetX() + field.middleRightGrid.getRegionWidth() / 2});
+    } else {
+        Vector2 enemy1 = enemyMap.begin()->second;
+        stpInfos["pass_defender_1"].setPositionToDefend(enemy1);
+        enemyMap.erase(enemyMap.begin());
+        Vector2 enemy2 = enemyMap.begin()->second;
+        stpInfos["pass_defender_2"].setPositionToDefend(enemy2);
+        enemyMap.erase(enemyMap.begin());
+        Vector2 enemy3 = enemyMap.begin()->second;
```
[the remark this answers is not in the document]
Comment (JornJorn) [reply roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp]: This is in an else statement only when they have at least 3 bots

File: roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp [hunk 5]
```diff
@@ -180,6 +152,73 @@ void AttackingPass::calculateInfoForAttackers() noexcept {
...
+
+        // Check if the pass_defenders are already assigned, if so check if the assignment is optimal
+        // and if not, update the assignment. This results in the pass_defenders not switching between enemies
+        // when another enemy is closer to the goal.
+        if (stpInfos["pass_defender_1"].getRobot() && stpInfos["pass_defender_2"].getRobot() && stpInfos["pass_defender_3"].getRobot()) {
+            std::vector<std::vector<double>> cost_matrix;
+            cost_matrix.resize(3);
+            for (int i = 0; i < 3; i++) {
+                cost_matrix[i].resize(3);
+            }
```
[the remark this answers is not in the document]
Comment (JornJorn) [reply roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp]: Thanks that looks way better, will take a look

File: roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp [hunk 6]
```diff
@@ -200,5 +239,5 @@ bool AttackingPass::shouldEndPlay() noexcept {
     return false;
 }
 
-const char* AttackingPass::getName() const { return "AttackingPass"; }
+const char* AttackingPass::getName() const { return "Attacking Pass"; }
```
[the remark this answers is not in the document]
Comment (JornJorn) [reply roboteam_ai/src/stp/plays/offensive/AttackingPass.cpp:127]: This was the only play without a space when there could be one :P Should we remove them all?

image

File: roboteam_ai/src/stp/plays/offensive/ChippingPass.cpp [hunk 7]
```diff
@@ -178,6 +150,73 @@ void ChippingPass::calculateInfoForAttackers() noexcept {
...
+
+        // Check if the pass_defenders are already assigned, if so check if the assignment is optimal
+        // and if not, update the assignment. This results in the pass_defenders not switching between enemies
+        // when another enemy is closer to the goal.
+        if (stpInfos["pass_defender_1"].getRobot() && stpInfos["pass_defender_2"].getRobot() && stpInfos["pass_defender_3"].getRobot()) {
+            std::vector<std::vector<double>> cost_matrix;
+            cost_matrix.resize(3);
+            for (int i = 0; i < 3; i++) {
+                cost_matrix[i].resize(3);
+            }
```
[the remark this answers is not in the document]
Comment (JornJorn) [reply roboteam_ai/src/stp/plays/offensive/ChippingPass.cpp]: Thanks that looks way better, will take a look

**Review by luukieboy — APPROVED** [review]

**Review by luukieboy — APPROVED** [review]

**Review by luukieboy — APPROVED** [review]

**Pull request merged**